apps/dashboard/src/components/workflow-editor/steps/base/data-object.tsx

```diff
@@ -0,0 +1,130 @@
+import { useMemo, useState } from 'react';
+import { FieldValues, useFormContext } from 'react-hook-form';
+
+import { Button } from '@/components/primitives/button';
+import { Card, CardContent } from '@/components/primitives/card';
+import { ControlInput } from '@/components/primitives/control-input';
+import { FormField } from '@/components/primitives/form/form';
+import { Input, InputRoot } from '@/components/primitives/input';
+import { useSaveForm } from '@/components/workflow-editor/steps/save-form-context';
+import { useWorkflow } from '@/components/workflow-editor/workflow-provider';
+import { parseStepVariablesToLiquidVariables } from '@/utils/parseStepVariablesToLiquidVariables';
+import { RiAddLine, RiDeleteBin6Line, RiInputField, RiQuestionLine } from 'react-icons/ri';
+import { Link } from 'react-router-dom';
+
+const dataObjectKey = 'data';
+
+const InnerDataObject = ({ field }: { field: FieldValues }) => {
+  const { saveForm } = useSaveForm();
+  const { step } = useWorkflow();
+
+  const variables = useMemo(() => (step ? parseStepVariablesToLiquidVariables(step.variables) : []), [step]);
+
+  const [currentPairs, setCurrentPairs] = useState(() => {
+    const obj = field.value ?? {};
+    return Object.entries(obj).map(([key, value]) => ({
+      key,
+      value: String(value ?? ''),
+    }));
+  });
+
+  const updateFormField = (pairs: Array<{ key: string; value: string }>) => {
+    setCurrentPairs(pairs);
+    const uniquePairLength = new Set<string>(pairs.map((pair) => pair.key)).size;
+    const hasNoDuplicates = uniquePairLength === pairs.length;
+
+    if (hasNoDuplicates) {
```

Comment:
If it has duplicates we need a validation error or we need to squash duplicates into a single value before saving.

Reply:
The reason I can't do anything better here is that we have an autosave mechanism. With the current approach, when a save is triggered (by blur), it will save the last valid field value that was recorded. We can't show an error because the form cannot be updated with duplicates, and duplicates usually form while typing as well.